src/main/java/htsjdk/samtools/cram/build/SliceFactory.java

```diff
@@ -0,0 +1,333 @@
+        for (final SAMRecord samRecord : samRecords) {
+            int referenceIndex = samRecord.getReferenceIndex();
+            final CRAMRecord cramRecord = new CRAMRecord(
+                    CramVersions.DEFAULT_CRAM_VERSION,
+                    encodingStrategy,
+                    samRecord,
+                    cramReferenceRegion.getReferenceBases(referenceIndex),
+                    recordIndex++,
+                    readGroupNameToID);
+            cramRecords.add(cramRecord);
+        }
+        resolveMatesForSlice(cramRecords);
+        return cramRecords;
+    }
+
+    //Update mate pairing
+    private void resolveMatesForSlice(final List<CRAMRecord> cramRecords) {
+        if (!coordinateSorted) {
+            for (final CRAMRecord r : cramRecords) {
+                r.setDetached(true);
+            }
+        } else {
+            //TODO: these maps are flawed in that while there should only be one primary record
+            // for a given read name, there can be multiple secondary reads, and they are not
+            // required to be encountered in order for coord-sorted, so its possible that we need
+            // to accumulate more than one read of a given name in the
+            final Map<String, CRAMRecord> primaryMateMap = new TreeMap<>();
+            final Map<String, CRAMRecord> secondaryMateMap = new TreeMap<>();
+            for (final CRAMRecord r : cramRecords) {
+                if (r.isReadPaired()) {
+                    final Map<String, CRAMRecord> mateMap =
+                            r.isSecondaryAlignment() ?
+                                    //TODO: is this right ?
+                                    primaryMateMap :
+                                    secondaryMateMap;
+                    CRAMRecord mate = mateMap.get(r.getReadName());
+                    mate = rejectBadMateChoice(mate, r);
+                    if (mate == null) {
+                        if (r.isSecondaryAlignment()) {
+                            secondaryMateMap.put(r.getReadName(), r);
+                        } else {
+                            primaryMateMap.put(r.getReadName(), r);
+                        }
+                    } else {
+                        mate.attachToMate(r);
+                    }
+                }
+            }
+
+            // mark unpredictable reads as detached:
+            for (final CRAMRecord cramRecord : cramRecords) {
+                cramRecord.updateDetachedState();
+            }
+        }
+    }
+
+    // We want to make sure that we choose the next mate in the fragment sequence to preserve
+    // the original order. If we just choose the first name that matches, it might not be the right one.
+    // So test the candidate mate to see if it matches the mate properties specified in the original read.
+    private CRAMRecord rejectBadMateChoice(final CRAMRecord firstMate, final CRAMRecord candidateMate) {
```

Comment:
What about templates with more than 2 reads? Don't we need to check that the mate is the next read in the sequence?

Reply:
(This one could be dealt with in a future branch)

Reply:
Code is removed.